internal/k8s/fake_client.go

```diff
@@ -120,7 +120,8 @@ func (c *FakeK8sClient) WatchEvents(ctx context.Context) (<-chan *v1.Event, erro
 }
 
 func (c *FakeK8sClient) EmitEvent(ctx context.Context, evt *v1.Event) {
-	c.eventsCh <- evt
+	// Do this async so it doesn't block if we haven't set up watch on this channel yet.
```

Comment:
have you considered setting up the eventsCh on startup?

i appreciate the temptation to fix race conditions with more async code, but it often just leads to even more race conditions.

Reply:
ooh ya that makes sense, will do

Reply:
oh! I remembered why I did it this way -- because if i create the eventsCh when we call `WatchEvts` it's easier to close it, i.e.:
```
go func() {
		<-ctx.Done()
		close(c.eventsCh)
	}()
```

I... have the sense that this is best practice but don't have a good sense of how necessary this is for tests. Thoughts?

Reply:
shouldn't we do this in TearDown?

Reply:
you mean adding a teardown func to the fake client? i'd be fine with that -- though seems kind of a pain for all the tests to remember to call it (i.e. for all the fixtures to remember to call it in their respective Teardown funcs). I'd kind of rather add a ctx to the call to `NewFakeClient`.

But that doesn't actually answer my question, which is: IS this sort of clean up in fact necessary in tests?

Reply:
(I guess I'm asking if you have objections to me putting that async channel-closing call in `NewFakeClient` only b/c you explicitly suggested putting it in `TearDown`)

Reply:
two separate points:

1) ya, you'll probably want to do this sort of cleanup in tests. it can be a pain to debug a test if there are a bunch of random goroutines from other tests hanging around

2) I worry that you're missing the high-level advice:

You added a goroutine to do teardown. 

Then you realized that this introduced a race condition, so you needed to add a new goroutine to handle the race condition. 

My high-level advice is that this is an understandable impulse, but likely to lead to even more race conditions. It might be better to consider how you could accomplish the goal with fewer goroutines

you can take or leave that advice

Reply:
right right, that's v useful advice. Thanks!